lib/amazonka/src/Amazonka/Auth/SSO.hs

```diff
@@ -0,0 +1,111 @@
+-- |
+-- Module      : Amazonka.Auth.SSO
+-- Copyright   : (c) 2013-2021 Brendan Hay
+-- License     : Mozilla Public License, v. 2.0.
+-- Maintainer  : Brendan Hay <brendan.g.hay+amazonka@gmail.com>
+-- Stability   : provisional
+-- Portability : non-portable (GHC extensions)
+module Amazonka.Auth.SSO where
+
+import Amazonka.Auth.Background (fetchAuthInBackground)
+import Amazonka.Auth.Exception
+import qualified Amazonka.Crypto as Crypto
+import Amazonka.Data.Sensitive
+import Amazonka.Data.Time (ISO8601, Time (..))
+import Amazonka.Env (Env, Env' (..))
+import Amazonka.HTTP (retryRequest)
+import Amazonka.Lens ((^.))
+import Amazonka.Prelude
+import Amazonka.SSO.GetRoleCredentials as SSO
+import qualified Amazonka.SSO.Types as SSO (RoleCredentials (..))
+import Amazonka.Types
+import qualified Control.Exception as Exception
+import Control.Exception.Lens (handling_, _IOException)
+import Control.Monad.Trans.Resource (runResourceT)
+import Data.Aeson (FromJSON, decodeFileStrict)
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+import Data.Time.Clock.POSIX (posixSecondsToUTCTime)
+import qualified Network.HTTP.Client as Client
+
+data Cache = Cache
+  { accessToken :: Text,
+    region :: Region
+  }
+  deriving stock (Generic)
+  deriving anyclass (FromJSON)
+
+fromSSO ::
+  (MonadIO m, Foldable withAuth) =>
+  FilePath ->
+  Region ->
+  Text ->
+  Text ->
+  Env' withAuth ->
+  m Env
+fromSSO cache ssoRegion accountId roleName env = do
+  auth <- liftIO $ fetchAuthInBackground getCredentials
+  pure $ env {envAuth = Identity auth}
+  where
+    getCredentials = do
+      Cache {..} <- readCache cache
+
+      let ssoEnv = env {envRegion = ssoRegion}
+
+      resp <-
+        runResourceT $
+          sendWithoutAuth ssoEnv $
+            SSO.newGetRoleCredentials roleName accountId accessToken
+
+      let mCreds = do
+            rc <- resp ^. SSO.getRoleCredentialsResponse_roleCredentials
+            roleCredentialsToAuthEnv rc
+
+      case mCreds of
+        Nothing -> fail "sso:GetRoleWithCredentials returned no credentials."
+        Just c -> pure c
+
+sendWithoutAuth ::
+  ( MonadResource m,
+    AWSRequest a,
+    Foldable withAuth
+  ) =>
+  Env' withAuth ->
+  a ->
+  m (AWSResponse a)
+sendWithoutAuth env req = do
+  eResponse <- retryRequest env req
+  either (liftIO . Exception.throwIO) (pure . Client.responseBody) eResponse
+
+relativeCacheFile :: MonadIO m => Text -> m FilePath
+relativeCacheFile startUrl = do
+  let sha1 = show $ Crypto.hashSHA1 $ Text.encodeUtf8 startUrl
+  pure $ "/.aws/sso/cache/" <> sha1 <> ".json"
```

Comment:
Can you please add a comment with that link to the `botocore` repo, and an explanation why we're doing this? Because so much of this behaviour is determined by reading source, it's better if we can make that easier for the devs who come after us. Example:

https://github.com/brendanhay/amazonka/blob/29e4888f847797533c968a855804990ecf8660dd/lib/amazonka-core/src/Amazonka/Request.hs#L182-L183